File: Sources/RepoBarCore/Models/RepositoryHost.swift
import Foundation

public struct RepositoryURLTemplates: Equatable, Codable, Sendable {
    public var repository: String?
    public var branch: String?
    public var commit: String?
    public var issue: String?
    public var pullRequest: String?

    public init(
        repository: String? = nil,
        branch: String? = nil,
        commit: String? = nil,
        issue: String? = nil,
        pullRequest: String? = nil
    ) {
        self.repository = repository
        self.branch = branch
        self.commit = commit
        self.issue = issue
        self.pullRequest = pullRequest
    }
}

public struct RepositoryHost: Identifiable, Equatable, Codable, Sendable {
    public var id: String
    public var provider: SourceControlProvider
    public var displayName: String
    public var webBaseURL: URL
    public var apiBaseURL: URL?
    public var authMethod: AuthMethod
    public var urlTemplates: RepositoryURLTemplates

    public init(
        id: String? = nil,
        provider: SourceControlProvider,
        displayName: String,
        webBaseURL: URL,
        apiBaseURL: URL?,
        authMethod: AuthMethod,
        urlTemplates: RepositoryURLTemplates = RepositoryURLTemplates()
    ) {
        self.provider = provider
        self.webBaseURL = webBaseURL
        self.id = id ?? Self.defaultID(provider: provider, webBaseURL: webBaseURL)
        self.displayName = displayName
        self.apiBaseURL = apiBaseURL
        self.authMethod = authMethod
        self.urlTemplates = urlTemplates
    }

    public static let githubCom = RepositoryHost(
        id: "github:github.com",
        provider: .github,
        displayName: "GitHub.com",
        webBaseURL: URL(string: "https://github.com")!,
        apiBaseURL: URL(string: "https://api.github.com")!,
        authMethod: .oauth
    )

    public static let gitlabCom = RepositoryHost(
        id: "gitlab:gitlab.com",
        provider: .gitlab,
        displayName: "GitLab.com",
        webBaseURL: URL(string: "https://gitlab.com")!,
        apiBaseURL: URL(string: "https://gitlab.com/api/v4")!,
        authMethod: .pat
    )

    public static let bitbucketCloud = RepositoryHost(
        id: "bitbucketCloud:bitbucket.org",
        provider: .bitbucketCloud,
        displayName: "Bitbucket Cloud",
        webBaseURL: URL(string: "https://bitbucket.org")!,
        apiBaseURL: URL(string: "https://api.bitbucket.org/2.0")!,
        authMethod: .apiToken
    )

    public static let codeberg = RepositoryHost(
        id: "forgejo:codeberg.org",
        provider: .forgejo,
        displayName: "Codeberg",
        webBaseURL: URL(string: "https://codeberg.org")!,
        apiBaseURL: URL(string: "https://codeberg.org/api/v1")!,
        authMethod: .pat
    )

    public static func defaultID(provider: SourceControlProvider, webBaseURL: URL) -> String {
        let host = webBaseURL.host?.lowercased() ?? webBaseURL.absoluteString.lowercased()
        return "\(provider.rawValue):\(host)"
    }
}

public struct RepositoryAccount: Equatable, Codable, Sendable {
    public var provider: SourceControlProvider
    public var webHost: URL
    public var apiHost: URL?
    public var username: String?

    public init(provider: SourceControlProvider, webHost: URL, apiHost: URL?, username: String? = nil) {
        self.provider = provider
        self.webHost = webHost
        self.apiHost = apiHost
        self.username = username
    }
}

File: Windows/RepoBar.Core/Models/RepositoryHost.cs
namespace RepoBar.Core.Models;

public sealed record RepositoryUrlTemplates(
    string? Repository = null,
    string? Branch = null,
    string? Commit = null,
    string? Issue = null,
    string? PullRequest = null);

public sealed record RepositoryHost(
    SourceControlProvider Provider,
    string DisplayName,
    Uri WebBaseUrl,
    Uri? ApiBaseUrl,
    AuthMethod AuthMethod,
    RepositoryUrlTemplates? UrlTemplates = null,
    string? Id = null)
{
    public string Id { get; init; } = Id ?? DefaultId(Provider, WebBaseUrl);

    public RepositoryUrlTemplates UrlTemplates { get; init; } = UrlTemplates ?? new RepositoryUrlTemplates();

    public static RepositoryHost GitHubCom { get; } = new(
        Provider: SourceControlProvider.GitHub,
        DisplayName: "GitHub.com",
        WebBaseUrl: new Uri("https://github.com"),
        ApiBaseUrl: new Uri("https://api.github.com"),
        AuthMethod: AuthMethod.OAuth,
        Id: "github:github.com");

    public static RepositoryHost GitLabCom { get; } = new(
        Provider: SourceControlProvider.GitLab,
        DisplayName: "GitLab.com",
        WebBaseUrl: new Uri("https://gitlab.com"),
        ApiBaseUrl: new Uri("https://gitlab.com/api/v4"),
        AuthMethod: AuthMethod.Pat,
        Id: "gitlab:gitlab.com");

    public static RepositoryHost BitbucketCloud { get; } = new(
        Provider: SourceControlProvider.BitbucketCloud,
        DisplayName: "Bitbucket Cloud",
        WebBaseUrl: new Uri("https://bitbucket.org"),
        ApiBaseUrl: new Uri("https://api.bitbucket.org/2.0"),
        AuthMethod: AuthMethod.ApiToken,
        Id: "bitbucketCloud:bitbucket.org");

    public static RepositoryHost Codeberg { get; } = new(
        Provider: SourceControlProvider.Forgejo,
        DisplayName: "Codeberg",
        WebBaseUrl: new Uri("https://codeberg.org"),
        ApiBaseUrl: new Uri("https://codeberg.org/api/v1"),
        AuthMethod: AuthMethod.Pat,
        Id: "forgejo:codeberg.org");

    public static string DefaultId(SourceControlProvider provider, Uri webBaseUrl) =>
        $"{ProviderId(provider)}:{webBaseUrl.Host.ToLowerInvariant()}";

    private static string ProviderId(SourceControlProvider provider) =>
        provider switch
        {
            SourceControlProvider.GitHub => "github",
            SourceControlProvider.GitLab => "gitlab",
            SourceControlProvider.BitbucketCloud => "bitbucketCloud",
            SourceControlProvider.Forgejo => "forgejo",
            SourceControlProvider.Gitea => "gitea",
            SourceControlProvider.CustomGit => "customGit",
            _ => provider.ToString(),
        };
}

public sealed record RepositoryAccount(
    SourceControlProvider Provider,
    Uri WebHost,
    Uri? ApiHost,
    string? Username = null);
